adapters/http-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -677,24 +672,15 @@ private void doUploadMessage(
                                 contentType,
                                 payload,
                                 props,
-                                currentSpan.context()),
-                        responseReady.future())
+                                currentSpan.context()).onFailure(thr -> responseReadyTracker.cancel("send event failed")),
+                        responseReadyTracker.waitForCommand())
                         .map(s -> (Void) null);
             }
         })
-        .compose(proceed -> {
+        .onSuccess(proceed -> {
             // overwrites the execution of closeHandler in setTtdRequestConnectionCloseHandler
             // to prevent ctx.response().end() from being called twice
             ctx.response().closeHandler(v -> logResponseGettingClosedPrematurely(ctx.getRoutingContext()));
-
-            // downstream message sent and (if ttd was set) command was received or ttd has timed out
-            // we wait for the CommandConsumer having been closed before delivering the response to the
-            // device in order to prevent a race condition when the device immediately sends a new
-            // request and the CommandConsumer from the current request has not been closed yet
-            return Optional.ofNullable(commandConsumerTracker.result())
-                    .map(consumer -> consumer.close(currentSpan.context())
-                            .otherwise(thr -> null))
-                    .orElseGet(Future::succeededFuture);
         })
         .map(proceed -> {
```

Comment:
Concerning the `if (ctx.response().closed()) {` block 2 lines below:
Sending of telemetry/event was actually successful here, therefore metrics should be reported as well.
(This is not really related to this PR, but could be added in this PR nonetheless, also because this code is now being invoked in more cases.)

Reply:
Fixed

Reply:
I don't see where it is fixed. I thought about having `metrics.reportTelemetry` be invoked in the block containing
"`failed to send http response for [{}] message from device [tenantId: {}, deviceId: {}]: response already closed`" on line 683. Or what did you mean here?

Reply:
ups. sorry. I thought about another comment.
Yes - nothing changed regarding that comment. Thought about leaving it for another PR.
But aster second look I think that metrics report from `else` should just be moved before `if/else`  to be executed unconditionally (because of forward success). Isn't so?

Reply:
reporting moved unconditionally before `if/else`

Reply:
I agree. The metrics having been reported in the `bodyEndHandler` before was probably due to the metric initially being named `incrementProcessedHttpMessages` (see [here](https://github.com/eclipse/hono/commit/c6df76d12844b4ae3a11acb66a9de2ea1b38406c#diff-8a5c90291d6bfbbcbd7ae69c43de9a5a5e77532137e677c23fd3e655545d67c5R599)), for which you could argue that only fully processed and answered HTTP requests count. With `reportTelemetry` this is different.
A subtle difference here is that the timer (given as `reportTelemetry` param) will now not include the duration for sending the response to the device. But I think this is not really relevant and is also not being taken into account when processing CoAP messages for example.